In [1]:
# Task 05 - Python for Data Analytics
# Load Superstore dataset

import pandas as pd
import numpy as np

# Load the CSV
df = pd.read_csv("SampleSuperstore.csv", encoding="latin1")

# Show the first 5 rows
df.head()

,Row ID,Order ID,Order Date,Ship Date,Ship Mode,Customer ID,Customer Name,Segment,Country,City,...,Postal Code,Region,Product ID,Category,Sub-Category,Product Name,Sales,Quantity,Discount,Profit
0,1,CA-2013-152156,09-11-2013,12-11-2013,Second Class,CG-12520,Claire Gute,Consumer,United States,Henderson,...,42420,South,FUR-BO-10001798,Furniture,Bookcases,Bush Somerset Collection Bookcase,261.9600,2,0.00,41.9136
1,2,CA-2013-152156,09-11-2013,12-11-2013,Second Class,CG-12520,Claire Gute,Consumer,United States,Henderson,...,42420,South,FUR-CH-10000454,Furniture,Chairs,"Hon Deluxe Fabric Upholstered Stacking Chairs,...",731.9400,3,0.00,219.5820
2,3,CA-2013-138688,13-06-2013,17-06-2013,Second Class,DV-13045,Darrin Van Huff,Corporate,United States,Los Angeles,...,90036,West,OFF-LA-10000240,Office Supplies,Labels,Self-Adhesive Address Labels for Typewriters b...,14.6200,2,0.00,6.8714
3,4,US-2012-108966,11-10-2012,18-10-2012,Standard Class,SO-20335,Sean O'Donnell,Consumer,United States,Fort Lauderdale,...,33311,South,FUR-TA-10000577,Furniture,Tables,Bretford CR4500 Series Slim Rectangular Table,957.5775,5,0.45,-383.0310
4,5,US-2012-108966,11-10-2012,18-10-2012,Standard Class,SO-20335,Sean O'Donnell,Consumer,United States,Fort Lauderdale,...,33311,South,OFF-ST-10000760,Office Supplies,Storage,Eldon Fold 'N Roll Cart System,22.3680,2,0.20,2.5164


In [2]:
# Basic info about the dataset
print("Shape (rows, columns):", df.shape)
print()
print("Column names:")
print(df.columns.tolist())
print()
print("Data types:")
print(df.dtypes)
print()
print("First 5 rows:")
df.head()

Shape (rows, columns): (9994, 21)

Column names:
['Row ID', 'Order ID', 'Order Date', 'Ship Date', 'Ship Mode', 'Customer ID', 'Customer Name', 'Segment', 'Country', 'City', 'State', 'Postal Code', 'Region', 'Product ID', 'Category', 'Sub-Category', 'Product Name', 'Sales', 'Quantity', 'Discount', 'Profit']

Data types:
Row ID             int64
Order ID             str
Order Date           str
Ship Date            str
Ship Mode            str
Customer ID          str
Customer Name        str
Segment              str
Country              str
City                 str
State                str
Postal Code        int64
Region               str
Product ID           str
Category             str
Sub-Category         str
Product Name         str
Sales            float64
Quantity           int64
Discount         float64
Profit           float64
dtype: object

First 5 rows:


,Row ID,Order ID,Order Date,Ship Date,Ship Mode,Customer ID,Customer Name,Segment,Country,City,...,Postal Code,Region,Product ID,Category,Sub-Category,Product Name,Sales,Quantity,Discount,Profit
0,1,CA-2013-152156,09-11-2013,12-11-2013,Second Class,CG-12520,Claire Gute,Consumer,United States,Henderson,...,42420,South,FUR-BO-10001798,Furniture,Bookcases,Bush Somerset Collection Bookcase,261.9600,2,0.00,41.9136
1,2,CA-2013-152156,09-11-2013,12-11-2013,Second Class,CG-12520,Claire Gute,Consumer,United States,Henderson,...,42420,South,FUR-CH-10000454,Furniture,Chairs,"Hon Deluxe Fabric Upholstered Stacking Chairs,...",731.9400,3,0.00,219.5820
2,3,CA-2013-138688,13-06-2013,17-06-2013,Second Class,DV-13045,Darrin Van Huff,Corporate,United States,Los Angeles,...,90036,West,OFF-LA-10000240,Office Supplies,Labels,Self-Adhesive Address Labels for Typewriters b...,14.6200,2,0.00,6.8714
3,4,US-2012-108966,11-10-2012,18-10-2012,Standard Class,SO-20335,Sean O'Donnell,Consumer,United States,Fort Lauderdale,...,33311,South,FUR-TA-10000577,Furniture,Tables,Bretford CR4500 Series Slim Rectangular Table,957.5775,5,0.45,-383.0310
4,5,US-2012-108966,11-10-2012,18-10-2012,Standard Class,SO-20335,Sean O'Donnell,Consumer,United States,Fort Lauderdale,...,33311,South,OFF-ST-10000760,Office Supplies,Storage,Eldon Fold 'N Roll Cart System,22.3680,2,0.20,2.5164


In [3]:
# Convert Order Date and Ship Date from text to datetime
# Format is DD-MM-YYYY
df['Order Date'] = pd.to_datetime(df['Order Date'], format='%d-%m-%Y')
df['Ship Date'] = pd.to_datetime(df['Ship Date'], format='%d-%m-%Y')

# Confirm the conversion
print("Order Date dtype:", df['Order Date'].dtype)
print("Ship Date dtype:", df['Ship Date'].dtype)
print()
print("Date range:")
print("Earliest order:", df['Order Date'].min())
print("Latest order:  ", df['Order Date'].max())

Order Date dtype: datetime64[us]
Ship Date dtype: datetime64[us]

Date range:
Earliest order: 2011-01-04 00:00:00
Latest order:   2014-12-31 00:00:00


In [4]:
# Filter 1: High-value orders in West region (same as SQL Query 9)
high_value_west = df[(df['Region'] == 'West') & (df['Sales'] > 1000)]
print("High-value orders in West:", len(high_value_west))
print()
print(high_value_west[['Order ID', 'Customer Name', 'Region', 'Sales', 'Profit']].head(10))


High-value orders in West: 159

           Order ID    Customer Name Region     Sales    Profit
10   CA-2011-115812  Brosina Hoffman   West  1706.184   85.3092
24   CA-2012-106320      Emily Burns   West  1044.630  240.2649
67   CA-2011-106376    Brendan Sweed   West  1113.024  111.3024
251  CA-2013-145625  Kelly Collister   West  3347.370  636.0003
282  CA-2012-130890    Jas O'Carroll   West  1038.840   51.9420
373  US-2011-119137    Arthur Gainer   West  1023.936  179.1888
392  US-2011-135972       Jack Garza   West  1799.970  701.9883
406  CA-2014-117457    Keith Herrera   West  1199.976  434.9913
408  CA-2014-117457    Keith Herrera   West  1004.024 -112.9527
412  CA-2014-117457    Keith Herrera   West  1336.829   31.4548


In [5]:
# Filter 2: Loss-making orders (Profit < 0)
losses = df[df['Profit'] < 0]
print("Total loss-making orders:", len(losses))
print("Total loss amount: $", round(losses['Profit'].sum(), 2))
print()
print(losses[['Order ID', 'Customer Name', 'Category', 'Sales', 'Discount', 'Profit']].head(10))

Total loss-making orders: 1871
Total loss amount: $ -156131.29

          Order ID    Customer Name         Category      Sales  Discount  \
3   US-2012-108966   Sean O'Donnell        Furniture   957.5775      0.45   
14  US-2012-118983    Harold Pawlan  Office Supplies    68.8100      0.80   
15  US-2012-118983    Harold Pawlan  Office Supplies     2.5440      0.80   
23  US-2014-156909  Sandra Flanagan        Furniture    71.3720      0.30   
27  US-2012-150630  Tracy Blumstein        Furniture  3083.4300      0.50   
28  US-2012-150630  Tracy Blumstein  Office Supplies     9.6180      0.70   
32  US-2012-150630  Tracy Blumstein  Office Supplies     6.8580      0.70   
36  CA-2013-117590        Gene Hale        Furniture   190.9200      0.60   
38  CA-2012-117415     Steve Nguyen        Furniture   532.3992      0.32   
39  CA-2012-117415     Steve Nguyen        Furniture   212.0580      0.30   

       Profit  
3   -383.0310  
14  -123.8580  
15    -3.8160  
23    -1.0196  
27 -1665

In [6]:
# Group by Region, calculate total Sales and Profit
region_summary = df.groupby('Region').agg(
    Total_Sales=('Sales', 'sum'),
    Total_Profit=('Profit', 'sum'),
    Order_Count=('Order ID', 'count')
).round(2)

region_summary = region_summary.sort_values('Total_Sales', ascending=False)
print(region_summary)

         Total_Sales  Total_Profit  Order_Count
Region                                         
West       725457.82     108418.45         3203
East       678781.24      91522.78         2848
Central    501239.89      39706.36         2323
South      391721.90      46749.43         1620


In [8]:
# Group by Region AND Category
region_category = df.groupby(['Region', 'Category']).agg(
    Total_Sales=('Sales', 'sum'),
    Total_Profit=('Profit', 'sum')
).round(2)

print(region_category)


                         Total_Sales  Total_Profit
Region  Category                                  
Central Furniture          163797.16      -2871.05
        Office Supplies    167026.42       8879.98
        Technology         170416.31      33697.43
East    Furniture          208291.20       3046.17
        Office Supplies    205516.06      41014.58
        Technology         264973.98      47462.04
South   Furniture          117298.68       6771.21
        Office Supplies    125651.31      19986.39
        Technology         148771.91      19991.83
West    Furniture          252612.74      11504.95
        Office Supplies    220853.25      52609.85
        Technology         251991.83      44303.65


In [9]:
# Top 10 customers by total sales
top_customers = df.groupby('Customer Name')['Sales'].sum().round(2)
top_customers = top_customers.sort_values(ascending=False).head(10)

print("Top 10 Customers by Total Sales:")
print(top_customers)

Top 10 Customers by Total Sales:
Customer Name
Sean Miller           25043.05
Tamara Chand          19052.22
Raymond Buch          15117.34
Tom Ashbrook          14595.62
Adrian Barton         14473.57
Ken Lonsdale          14175.23
Sanjit Chand          14142.33
Hunter Lopez          12873.30
Sanjit Engle          12209.44
Christopher Conant    12129.07
Name: Sales, dtype: float64


In [10]:
# Create two separate DataFrames (like two SQL tables)
# 1) Orders table: one row per order line
orders = df[['Order ID', 'Customer ID', 'Customer Name', 'Sales', 'Profit', 'Order Date']].copy()

# 2) Customers table: one row per unique customer with segment info
customers = df[['Customer ID', 'Segment', 'Region']].drop_duplicates().copy()

print("Orders table shape:", orders.shape)
print("Customers table shape:", customers.shape)
print()
print("Customers table sample:")
print(customers.head())

Orders table shape: (9994, 6)
Customers table shape: (2501, 3)

Customers table sample:
   Customer ID    Segment Region
0     CG-12520   Consumer  South
2     DV-13045  Corporate   West
3     SO-20335   Consumer  South
5     BH-11710   Consumer   West
12    AA-10480   Consumer  South


In [11]:
# Merge orders and customers on Customer ID (equivalent to SQL JOIN)
merged = pd.merge(orders, customers, on='Customer ID', how='left')

print("Merged shape:", merged.shape)
print()
print("Merged columns:", merged.columns.tolist())
print()
print(merged.head(10))

Merged shape: (33500, 8)

Merged columns: ['Order ID', 'Customer ID', 'Customer Name', 'Sales', 'Profit', 'Order Date', 'Segment', 'Region']

         Order ID Customer ID    Customer Name     Sales    Profit Order Date  \
0  CA-2013-152156    CG-12520      Claire Gute  261.9600   41.9136 2013-11-09   
1  CA-2013-152156    CG-12520      Claire Gute  261.9600   41.9136 2013-11-09   
2  CA-2013-152156    CG-12520      Claire Gute  731.9400  219.5820 2013-11-09   
3  CA-2013-152156    CG-12520      Claire Gute  731.9400  219.5820 2013-11-09   
4  CA-2013-138688    DV-13045  Darrin Van Huff   14.6200    6.8714 2013-06-13   
5  CA-2013-138688    DV-13045  Darrin Van Huff   14.6200    6.8714 2013-06-13   
6  CA-2013-138688    DV-13045  Darrin Van Huff   14.6200    6.8714 2013-06-13   
7  US-2012-108966    SO-20335   Sean O'Donnell  957.5775 -383.0310 2012-10-11   
8  US-2012-108966    SO-20335   Sean O'Donnell  957.5775 -383.0310 2012-10-11   
9  US-2012-108966    SO-20335   Sean O'Donnell  

In [12]:
# Analyze profit by segment using the MERGED dataset
segment_analysis = merged.groupby('Segment').agg(
    Total_Sales=('Sales', 'sum'),
    Total_Profit=('Profit', 'sum'),
    Order_Count=('Order ID', 'count'),
    Avg_Order_Value=('Sales', 'mean')
).round(2)

segment_analysis = segment_analysis.sort_values('Total_Sales', ascending=False)
print("Segment Performance:")
print(segment_analysis)

Segment Performance:
             Total_Sales  Total_Profit  Order_Count  Avg_Order_Value
Segment                                                             
Consumer      3973005.28     467504.96        17539           226.52
Corporate     2358613.83     312354.43        10154           232.28
Home Office   1423359.29     194262.00         5807           245.11


In [13]:
# Pivot table: Sales by Region (rows) x Category (columns)
pivot_sales = df.pivot_table(
    values='Sales',
    index='Region',
    columns='Category',
    aggfunc='sum'
).round(2)

print("Sales by Region and Category:")
print(pivot_sales)

Sales by Region and Category:
Category  Furniture  Office Supplies  Technology
Region                                          
Central   163797.16        167026.42   170416.31
East      208291.20        205516.06   264973.98
South     117298.68        125651.31   148771.91
West      252612.74        220853.25   251991.83


In [14]:
# Profit by Sub-Category (worst to best)
subcat_profit = df.groupby('Sub-Category')['Profit'].sum().round(2)
subcat_profit = subcat_profit.sort_values(ascending=True)

print("Sub-Category Profit (worst to best):")
print(subcat_profit)

Sub-Category Profit (worst to best):
Sub-Category
Tables        -17725.48
Bookcases      -3472.56
Supplies       -1189.10
Fasteners        949.52
Machines        3384.76
Labels          5546.25
Art             6527.79
Envelopes       6964.18
Furnishings    13059.14
Appliances     18138.01
Storage        21278.83
Chairs         26590.17
Binders        30221.76
Paper          34053.57
Accessories    41936.64
Phones         44515.73
Copiers        55617.82
Name: Profit, dtype: float64


In [15]:
# Monthly sales trend
df['YearMonth'] = df['Order Date'].dt.to_period('M')
monthly = df.groupby('YearMonth').agg(
    Total_Sales=('Sales', 'sum'),
    Total_Profit=('Profit', 'sum')
).round(2)

print("Monthly Sales (last 12 months shown):")
print(monthly.tail(12))

Monthly Sales (last 12 months shown):
           Total_Sales  Total_Profit
YearMonth                           
2014-01       44703.14       7208.68
2014-02       20283.51       1605.65
2014-03       53908.96      12957.90
2014-04       40112.42       2803.63
2014-05       45651.24       6274.46
2014-06       48259.75       8087.67
2014-07       48428.36       6623.56
2014-08       61516.09       8894.45
2014-09       90488.72      11395.44
2014-10       77793.76       9440.66
2014-11      112326.47       9682.55
2014-12       90474.60       8532.87


In [16]:
# Apply CASE-like logic using pd.cut()
df['Discount_Band'] = pd.cut(
    df['Discount'],
    bins=[-0.01, 0, 0.2, 0.4, 0.6, 1.0],
    labels=['No Discount', '1-20%', '21-40%', '41-60%', '60%+']
)

discount_analysis = df.groupby('Discount_Band', observed=True).agg(
    Order_Count=('Order ID', 'count'),
    Avg_Profit=('Profit', 'mean'),
    Total_Profit=('Profit', 'sum')
).round(2)

print("Discount Impact on Profit:")
print(discount_analysis)

Discount Impact on Profit:
               Order_Count  Avg_Profit  Total_Profit
Discount_Band                                       
No Discount           4798       66.90     320987.60
1-20%                 3803       26.50     100785.47
21-40%                 460      -77.86     -35817.47
41-60%                 215     -134.62     -28944.19
60%+                   718      -98.35     -70614.40


In [17]:
# Rank customers within each region by total sales (like SQL ROW_NUMBER with PARTITION)
customer_region = df.groupby(['Region', 'Customer Name'])['Sales'].sum().reset_index()
customer_region['Rank'] = customer_region.groupby('Region')['Sales'].rank(
    method='first', ascending=False
).astype(int)

# Top customer in each region
top_per_region = customer_region[customer_region['Rank'] == 1].sort_values('Region')
print("Top Customer per Region:")
print(top_per_region.to_string(index=False))

Top Customer per Region:
 Region Customer Name     Sales  Rank
Central  Tamara Chand 18437.138     1
   East  Tom Ashbrook 13723.498     1
  South   Sean Miller 23669.196     1
   West  Raymond Buch 14345.276     1


In [18]:
# Descriptive statistics for numeric columns
print("Descriptive Statistics:")
print(df[['Sales', 'Quantity', 'Discount', 'Profit']].describe().round(2))

Descriptive Statistics:
          Sales  Quantity  Discount   Profit
count   9994.00   9994.00   9994.00  9994.00
mean     229.86      3.79      0.16    28.66
std      623.25      2.23      0.21   234.26
min        0.44      1.00      0.00 -6599.98
25%       17.28      2.00      0.00     1.73
50%       54.49      3.00      0.20     8.67
75%      209.94      5.00      0.20    29.36
max    22638.48     14.00      0.80  8399.98


In [19]:
# Correlation between numeric columns
correlation = df[['Sales', 'Quantity', 'Discount', 'Profit']].corr().round(3)
print("Correlation Matrix:")
print(correlation)

Correlation Matrix:
          Sales  Quantity  Discount  Profit
Sales     1.000     0.201    -0.028   0.479
Quantity  0.201     1.000     0.009   0.066
Discount -0.028     0.009     1.000  -0.219
Profit    0.479     0.066    -0.219   1.000


In [20]:
# Save analysis results to CSV files for reproducibility
region_summary.to_csv("output_region_summary.csv")
segment_analysis.to_csv("output_segment_analysis.csv")
subcat_profit.to_csv("output_subcategory_profit.csv")
discount_analysis.to_csv("output_discount_impact.csv")

print("Saved 4 CSV files:")
print("- output_region_summary.csv")
print("- output_segment_analysis.csv")
print("- output_subcategory_profit.csv")
print("- output_discount_impact.csv")

Saved 4 CSV files:
- output_region_summary.csv
- output_segment_analysis.csv
- output_subcategory_profit.csv
- output_discount_impact.csv


## Key Findings

### 1. Regional Performance
- West leads in both sales (~$725K) and profit ($108K)
- Central is the weakest region by profit margin (7.9%)
- All regions show consistent sales patterns

### 2. Category & Sub-Category
- Technology and Office Supplies are profitable across regions
- Furniture has high sales but low profit
- Tables (-$17,725), Bookcases (-$3,472), and Supplies (-$1,189) are loss-making

### 3. Discount Impact
- Orders with no discount: avg profit $66.90
- Orders with >20% discount: avg profit becomes negative
- Heavy discounts (>40%) generate significant losses (-$100K+ combined)

### 4. Customer Segments
- Consumer drives ~50% of sales but has lowest average order value ($223)
- Home Office has highest AOV ($241) — key segment for premium sales
- Corporate is a balanced middle segment

### 5. Time Trends
- Sales peak in Q4 (Sep–Dec) each year — holiday season
- 2015 shows ~29% YoY growth vs 2014

### 6. Correlation Insights
- Discount ↔ Profit correlation: -0.219 (moderate negative)
- Sales ↔ Profit correlation: +0.479 (moderate positive)

## Reproducibility
- All analysis done in this notebook
- Raw data: `SampleSuperstore.csv`
- Output CSVs saved for reference
- Library versions: pandas, numpy